src/codegen/compile.ml

```diff
@@ -7221,9 +7243,25 @@ module FuncDec = struct
       add_cycles ^^
       (* done! *)
       IC.system_call env "call_perform" ^^
-      (* Check error code *)
+      IC.set_call_perform_status env ^^
+      Blob.lit env message ^^
+      IC.set_call_perform_message env ^^
+      IC.get_call_perform_status env ^^
+      (* save error code, cleanup on error *)
       G.i (Test (Wasm.Values.I32 I32Op.Eqz)) ^^
-      E.else_trap_with env (Printf.sprintf "could not perform %s" purpose)
+      G.if0
```

Comment:
I am a bit confused by this. `status` = 0 means no error. `Test` will turn that around to `1`. Then `if0` will proceed to the _else_ branch, cleaning up...

My feeling is that we should be able to eliminate the `Test` instruction.

Reply:
Agree, fixing in 3727